src/lib.rs

```diff
@@ -248,12 +249,52 @@ impl<T> ThinVec<T> {
         }
     }
 
+    /// Reserve capacity for at least `additional` more elements to be inserted.
+    ///
+    /// May reserve more space than requested, to avoid frequent reallocations.
+    ///
+    /// Panics if the new capacity overflows `usize`.
+    ///
+    /// Re-allocates only if `self.capacity() < self.len() + additional`.
     pub fn reserve(&mut self, additional: usize) {
-        // TODO
+        let len = self.len();
+        let old_cap = self.capacity();
+
+        // Ensure the new capacity is at least double, to guarantee exponential growth.
+        let min_cap = len.checked_add(additional).expect("capacity overflow");
+        let double_cap = old_cap.saturating_mul(2);
+        let new_cap = max(min_cap, double_cap);
+
+        let additional = new_cap - len;
+        self.reserve_exact(additional);
```

Comment:
Hopefully the redundant checks get optimized out; otherwise we could just copy/paste more code here.